index.bs

```diff
@@ -314,12 +321,40 @@ See <a href="https://lists.w3.org/Archives/Public/public-speech-api/2012Sep/0072
+
+    When invoked, run these steps:
+    1. Let <var>promise</var> be <a>a new promise</a>.
+    1. Run the <a>on-device availability algorithm</a> with <var>lang</var>. If it returns an exception, throw it and abort these steps.
+    1. Return <var>promise</var> and resolve it with the result of the <a>on-device availability algorithm</a>.
+  </dd>
 
   <dt><dfn method for=SpeechRecognition>installOnDevice({{DOMString}} lang)</dfn> method</dt>
-  <dd>The installOnDevice method returns a Promise that resolves to a boolean indicating whether the installation of on-device speech recognition for a given BCP 47 language tag initiated successfully. [[!BCP47]]</dd>
+  <dd>
+    The {{SpeechRecognition/installOnDevice}} method returns a {{Promise}} that resolves to a {{boolean}} indicating whether the installation of on-device speech recognition for a given [[!BCP47]] language tag initiated successfully.
+
+    When invoked, run these steps:
+    1. Let <var>promise</var> be <a>a new promise</a>.
+    1. Run the <a>on-device availability algorithm</a> with <var>lang</var>. If it returns an exception, throw it and abort these steps.
+    1. Return <var>promise</var>.
+    1. If the <a>on-device availability algorithm</a> returns {{AvailabilityStatus/unavailable}}, {{AvailabilityStatus/downloading}}, or {{AvailabilityStatus/available}}, resolve <var>promise</var> with <code>false</code> and skip the rest of these steps.
+    1. Initiate the download of the on-device speech recognition language for <var>lang</var>.
+        <p class=note>
+          Note: The user agent can prompt the user for explicit permission to download the on-device speech recognition language pack.
+        </p>
+    1. If the initiation of the download succeeds, resolve <var>promise</var> with <code>true</code>, otherwise resolve it with <code>false</code>.
+  </dd>
 
 </dl>
+<p>When the  <dfn>on-device availability algorithm</dfn> with <var>lang</var> is invoked, the user agent MUST run the following steps:
+1. If the [=current settings object=]'s [=relevant global object=]'s [=associated Document=] is NOT [=fully active=], throw an {{InvalidStateError}} and abort these steps.
+1. If <var>lang</var> is not a valid [[!BCP47]] language tag, throw a {{SyntaxError}} and abort these steps.
+1. In parallel, do the following steps:
+    - If the on-device speech recognition language pack for <var>lang</var> is unsupported, return {{AvailabilityStatus/unavailable}}.
+    - If the on-device speech recognition language pack for <var>lang</var> is supported but not installed, return {{AvailabilityStatus/downloadable}}.
+    - If the on-device speech recognition language pack for <var>lang</var> is downloading, return {{AvailabilityStatus/downloading}}.
+    - If the on-device speech recognition language pack for <var>lang</var> is installed, return {{AvailabilityStatus/available}}.
```

Comment:
We need to add steps to queue a task back to the main thread here, with the value decided by the four lines. In this task, we can resolve the promise with this value.

Because of this, this algorithm needs to take an additional parameter, the promise. The queue in which the task is enqueued is inferred by the fact that we're checking on an `associated Document` etc., so generally we're not too specific.